### PR TITLE
chore: implement pausable in ServiceManager

### LINKED ISSUE #480: chore: implement pausable in ServiceManager

### PR DESCRIPTION
**Pull request by JuArce**
# How to test
0. Addresses to export
Note:
```
export RPC_URL=http://localhost:8545
export ALIGNED_SERVICE_MANAGER=0x851356ae760d987E095750cCeb3bC6014560891C
export BATCHER_PAYMENT_SERVICE=0x7bc06c482DEAd17c0e297aFbC32f6e63d3846650
export ALIGNED_SERVICE_MANAGER_PAUSER_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
export BATCHER_PAYMENT_SERVICE_PAUSER_PRIVATE_KEY=0x4bbbf85ce3377467afe5d46f804f221813b2bb87f24d81f60f1fcdbf7cbf4356
```
2. Run anvil
```
make anvil_start_with_block_time
```
3. If you run 
```
make get_paused_state_aligned_service_manager
```
the response should be `No functions are paused`

4. If you run
```
make pause_all_aligned_service_manager
```
You have paused the whole aligned layer service manager contract.
Thus if you run 
```
make get_paused_state_aligned_service_manager
```
the response should be `All functions are paused`
5. You can also run 
```
make unpause_all_aligned_service_manager
```
To unpause the whole contract.
6. You can also run the following command, to pause or unpause specific functions, receiving a list of the functions to remain paused (this is how the Pausable parameter works in the contract).
For example, if you want to pause functions 0, 2 and 3, you can run
```
contracts/scripts/pause_aligned_service_manager.sh 0 2 3
```
Then, if you want to unpause, say, function 2, you must run 
```
contracts/scripts/unpause_aligned_service_manager.sh 0 3
```

Note: when executing a Pause, you can only ADD functions to the paused list, and when executing an Unpause, you can only REMOVE functions from the paused list. This is because the base pausable contract has different ACL for Pausers and Unpausers.

7. Then, you can experiment Pausing and Unpausing different functions, and running different parts of the system (for example, try sending a proof with the createNewTask function paused, or send one with only the respondToTask function paused)

Note: the list of pausable functions and their numbers can be seen in the `AlignedLayerServiceManager.sol` contract. But the list is the following:

0. createNewTask
1. respondToTaskV2
2. verifyBatchInclusion
3. withdraw
4. depositToBatcher
5. receive

---

BatcherPayments is also pausable, but without so much detail. You can either pause or unpause the contract running the following:
```
make pause_batcher_payment_service
make unpause_batcher_payment_service
```

And this will either pause or unpause the following functions:
- createNewTask
- unlock
- lock
- withdraw

# How to Deploy
## New Deployments
For new deployments, just run 
```
make deploy_aligned_contracts
```
PauserRegistry will be deployed and ServiceManager will be pausable

## Existing deployments
1. In order to use the EigenLayer pausable pattern with PauserRegistry contract, we have to deploy the PauserRegistry contract
```
make deploy_pauser_registry
```
Once deployed, update the output file with the "pauserRegistry" and the "alignedLayerPauser"

2. Run 
```
make upgrade_pauser_aligned_contracts
```
This will deploy a new ServiceManager with Pausable and will upgrade the Proxy. Also, it will run initializePaused()

### CODE REVIEW

File: contracts/script/deploy/AlignedLayerDeployer.s.sol
Comment (entropidelic) [inline contracts/script/deploy/AlignedLayerDeployer.s.sol]: just wondering if this is still needed, due to the TODO comment

Comment (JuArce) [reply contracts/script/deploy/AlignedLayerDeployer.s.sol]: > just wondering if this is still needed, due to the TODO comment

I left this comment because the deployer is setting the EigenLayer address as Pauser of our contracts. I want to fix this bug in another PR

Comment (JuArce) [reply contracts/script/deploy/AlignedLayerDeployer.s.sol]: This PR solves this TODO https://github.com/yetanotherco/aligned_layer/pull/501

**Review by entropidelic — APPROVED** [review]

File: contracts/src/core/AlignedLayerServiceManager.sol
Comment (entropidelic) [inline contracts/src/core/AlignedLayerServiceManager.sol]:
```suggestion
    // Once the contract is pausable this method is not needed anymore
```

**Review by uri-99 — APPROVED** [review]

**Comment by uri-99** [comment]
I can't be a reviewer because I made big changes on it

**Review by uri-99 — APPROVED** [review]
I can't be a reviewer because I made big changes on it

**Review by MarcosNicolau — APPROVED** [review]
I tested this on my machine and everything worked fine. Trying to use a paused function indeed reverts the transaction.

File: contracts/scripts/pause_aligned_service_manager.sh
Comment (Oppen) [inline contracts/scripts/pause_aligned_service_manager.sh]: `return` is meant to exit from a function call, not a script. You should replace this with `exit 0`.

**Comment by Oppen** [comment]
@entropidelic could you do another round on this? I don't know enough Solidity to trust myself on reviewing this and Uri is tainted by being the author of a big part of the PR.

File: contracts/scripts/pause_aligned_service_manager.sh
Comment (uri-99) [reply contracts/scripts/pause_aligned_service_manager.sh]: refactored

**Review by uri-99 — CHANGES_REQUESTED** [review]
blocking until i make a short how-to-use tutorial on pausables

**Review by uri-99 — APPROVED** [review]

**Comment by uri-99** [comment]
waiting for a review of @JuArce before merge

**Review by JuArce — COMMENTED** [review]
I think BatcherPaymentService should have the logic to pause/unpase functions individually

**Review by uri-99 — CHANGES_REQUESTED** [review]
blocking until batcherPayments is is pausable individually

**Comment by uri-99** [comment]
> I think BatcherPaymentService should have the logic to pause/unpase functions individually

I agree, i have created the [issue](https://github.com/yetanotherco/aligned_layer/issues/1271) and [started the PR](https://github.com/yetanotherco/aligned_layer/pull/1272) of this functionality.

Mauro has given thumbs up that the initial pausable version can not contain this, as it is Pausable as a whole. 

Also, this PR is specifically of adding pausable to ServiceManager, and this feature is non trivial, with probably many file changes.

**Review by uri-99 — APPROVED** [review]
Unblock since Batcher individual pausing will be added in [another PR](https://github.com/yetanotherco/aligned_layer/pull/1272)

**Comment by Oppen** [comment]
> Unblock since Batcher individual pausing will be added in [another PR](https://github.com/yetanotherco/aligned_layer/pull/1272)

I'd suggest to block this until #1272 is merged

**Review by JuArce — COMMENTED** [review]
LGTM

**Review by MarcosNicolau — APPROVED** [review]
Re-tested with the merge and it works fine. As a note, we need to make sure we update the upgrade for the multisig later in #516.

**Pull request merged**
